pkg/controllers/node/kdd.go

```diff
@@ -0,0 +1,199 @@
+		for _, idx := range b.Allocations {
+			if idx == nil {
+				// Not allocated.
+				continue
+			}
+			attr := b.Attributes[*idx]
+
+			// Track nodes based on IP allocations.
+			if val, ok := attr.AttrSecondary["node"]; ok {
+				if _, ok := nodes[val]; !ok {
+					nodes[val] = []allocation{}
+				}
+
+				// If there is no handle, then skip this IP. We need the handle
+				// in order to release the IP below.
+				if attr.AttrPrimary == nil {
+					log.Warnf("Skipping IP with no handle")
+					continue
+				}
+
+				// Add this allocation to the node, so we can release it later if
+				// we need to.
+				nodes[val] = append(nodes[val], allocation{
+					handle:     *attr.AttrPrimary,
+					attributes: attr.AttrSecondary,
+				})
+			}
+		}
+	}
+	log.Debugf("Nodes in IPAM: %s", nodes)
+
+	// For each node present in IPAM, if it doesn't exist in the Kubernetes API then we
+	// should consider it a candidate for cleanup.
+	for node, allocations := range nodes {
+		// Check if it exists in the Kubernetes API.
+		logc := log.WithField("node", node)
+		if c.nodeExists(node) {
+			logc.Debug("Node still exists, continue")
+			continue
+		}
+
+		// Node exists in IPAM but not in the Kubernetes API. Go through each IP address and
+		// check to see if the pod it references exists. If all the pods on that node are gone,
+		// continue with deletion. If any pod still exists, we skip this node. We want to be
+		// extra sure that the node is gone before we clean it up.
+		canDelete := true
+		for _, a := range allocations {
+			ns := a.attributes["namespace"]
+			pod := a.attributes["pod"]
+
+			// TODO: Need to handle IPIP addresses somehow. They are allocated in calico-ipam, but
+			// currently don't have this metadata attached.
+			if ns == "" || pod == "" {
+				log.Warnf("IP allocation does not have a namespace/pod associated")
+				continue
+			}
+
+			// Check to see if the pod still exists. If it does, then we shouldn't clean up
+			// this node, since it might come back online.
+			if c.podExists(pod, ns) {
+				logc.WithFields(log.Fields{"pod": pod, "ns": ns}).Debugf("Pod still exists")
+				canDelete = false
+				break
+			}
+		}
+
+		if !canDelete {
+			// Return an error here, it will trigger a reschedule of this call.
+			logc.Infof("Can't cleanup node yet - pods still exist")
+			return fmt.Errorf("Cannot clean up node yet, pods still exist")
+		}
+
+		logc.Info("Cleaning up IPAM resources for deleted node")
+		if err := c.cleanupNode(node, allocations); err != nil {
+			// TODO: We might want to clean up other nodes before returning.
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *NodeController) cleanupNode(node string, allocations []allocation) error {
+	// Potentially ratelimit node cleanup.
+	time.Sleep(c.rl.When(RateLimitCalicoDelete))
+
+	// Release the affinities for this node.
+	logc := log.WithField("node", node)
+	if err := c.calicoClient.IPAM().ReleaseHostAffinities(c.ctx, node); err != nil {
```

Comment:
I'm a little bit worried about the case where a node is considered "deleted" above, but then is recreated before we execute this line.

In that scenario, I think we'll release any affine blocks which may have been allocated to the "new" version of that node.

It's still recoverable I think - it just means we might end up with a block with no affinity, which will be released once all the IPs are freed (either when the pods are torn down or the node is eventually deleted again).